### PR TITLE
[parameters] Save parameters to file, Relation constraint

### PR DESCRIPTION
**Pull request by dgramop**
Edit parameterized constraints in the GUI, (instead of their computed values) and save those constraints to the file.

Continues work on #77 on the parameters branch

With these changes, solvespace can still open files created by older versions. Older versions cannot open files created by this version, however.

### CODE REVIEW

**Comment by phkahler** [comment]
> Instead of working on this completely solo and then dumping a big PR with all my progress, figured I should check in with some small changes to get your feedback and make sure I'm on the right track

@dgramop Thanks. I really appreciate you getting into this.  I'll look at what I can, but I've been a bit detached from Solvespace lately. @ruevs seems to be in a similar state ;-) but he might review some stuff too. And of course others pop in sometimes totally unexpected with feedback!

The approach of extending the parameters branch is probably good. I'll have a look at this this weekend.

**Comment by dgramop** [comment]
Based on our discussion in #77, we may end up overhauling the dimension system anyhow. I can clean up the scale factor solution currently in my PR so that we have _something_ to play around with until the unit system can be overhauled (if that's what is decided in #77).

**Comment by dgramop** [comment]
Alright looks like everyone's good to do the scaling solution. I'm going to still mark this as WIP since there are few more changes that I'd like to add on now that I've tested this version a little more

**Comment by dgramop** [comment]
Still in progress.... 

Uploading Screen Recording 2024-04-06 at 4.14.42 AM.mov…

Here is a proof of concept!

**Comment by dgramop** [comment]
The following is a snapshot of what still needs to be done before this is ready for review: 

- [ ] **Correctly Update DOF calculations when finished editing relation, since variables mentioned could have changed**
	- Currently if you create a constraint that doesn't reduce DOF, DOF still goes down by one until you start dragging stuff (which seems to force an update)
- [ ] **Variable Scoping**
	- Currently, to reference a variable in a sketch, for example in another group like an extrusion, I have to make another dimension around something that reflects that variable and give it a variable in the new scope to control 
- [ ] **General code quality cleanup**
	- [ ] Fix up the scaling factors
		- [ ] Make sure scaling factors are never applied on RELATIONs
		- [ ] Simplify expressions with duplicate scaling factors
	- [ ] Rebase and squash
- [ ] **Testing & error handling**
	- [ ] Erroneous input
		- [ ] x = 2x
		- [ ] x = y, y= 2x

**Comment by dgramop** [comment]
I decided to postpone variable scoping into a later PR. I think it's best to merge changes into our `parameters` branch since these changes have reached atomicity.

The next PR into that branch will be variable scoping, by which point we should (IMO) have more than enough to start discussing merging to `parameters` to master.

**Comment by dgramop** [comment]
Maintainers may notice I cut some scope from this PR, felt it was better to chunk this up than "slip" any more changes through in this one PR.

File: src/solvespace.h [hunk 1]
```diff
@@ -621,7 +621,7 @@ class SolveSpaceUI {
     std::string MmToString(double v, bool editable=false);
     std::string MmToStringSI(double v, int dim = 0);
     std::string DegreeToString(double v);
-    double ExprToMm(Expr *e);
+    double NonConstraintExprToMm(Expr *e);
```
Comment (dgramop) [inline src/solvespace.h:624]: I chose to rename this function, since using it to get the value of a constraint's expression in base Mm is a bad idea, considering scale factor stapled to each expression in a constraint, which encodes the scale factor that constraint was "saved" under, which itself was necessarily to correctly render constraint labels while applying the right scalings. 

By renaming the function, I'm hoping to force callers to declare compliance with this invariant

File: res/icons/graphics-window/relation.png
Comment (dgramop) [inline res/icons/graphics-window/relation.png:1]: Picasso, I consider myself

**Comment by ruevs** [comment]
@dgramop I've been following your progress with great interest. Unfortunately only from a phone at odd times of day when I had some spare time.

I'll try to spend some time on my SolveSpace development system and will pull, review and try your changes.

I agree I should merge them into "parameters" first (no squashing so that we have the history of how things happened) - this way you will have a new "base" for further changes. When we decide that it is time to merge to master I'll do the rabasing and squashing (to a few "meaningful" commits).

File: res/icons/graphics-window/relation.png
Comment (ruevs) [reply res/icons/graphics-window/relation.png:1]: That is a beautiful `=` sign :-) ;-)

**Comment by dgramop** [comment]
Nice! By squash I meant my own parameters_messy branch, just so the PR is a single commit. Looking forward to more parametric features in solvespace!

**Comment by dgramop** [comment]
@ruevs just rewrote history so hard authoritarian regimes are taking notice. in any case, I went ahead and undid my squash and got pushed the full history in all its glory.

**Comment by ruevs** [comment]
> @ruevs just rewrote history so hard authoritarian regimes are taking notice.

Just for context so others know what we are talking about https://github.com/dgramop/solvespace/pull/1 ;-)

@dgramop I played with your branch a bit. It actually works surprisingly well. Impressively thorough job!

Here is my first file using parameters: [first_dgramop_parameters_test.zip](https://github.com/solvespace/solvespace/files/14994577/first_dgramop_parameters_test.zip)
image

I'm particularly impressed and (pleasantly) surprised that there can be two "distance" constraints on the same line segment when one is an "equation". This is beautiful and I'll definitely dig into the code until I figure out why it is possible. I did also notice some interesting corner cases but let me first go through the code more carefully and add "formal" review comments here.

Another "sad" finding is that opening the above file in previous versions causes them to "crash" after the:

```
---------------------------
SolveSpace - Error
---------------------------
Unrecognized data in file.

This file may be corrupt, or from a newer version of the program.
---------------------------
OK   
---------------------------
```

dialog.

- 1.9, 2.0, 2.1, 2.3  close with no explanation.
- 3.0.rc1, 3.0, 3.1 crash an show the proper diagnostic:

```
---------------------------
Fatal error â€” SolveSpace
---------------------------
File D:\a\solvespace\solvespace\src\constrainteq.cpp, line 1057, function SolveSpace::ConstraintBase::GenerateEquations:
Assertion failed: false.
Message: Unexpected constraint ID.

Generate debug report?
---------------------------
OK   Cancel   
---------------------------
```

Of course this is [normal and "unavoidable"](https://github.com/solvespace/solvespace/blob/7d379e7f3b045ae579537abd392d5b368a4117e5/src/constrainteq.cpp#L1057) considering the new [`Type::RELATION` constraint type](https://github.com/dgramop/solvespace/blob/10eafdaf1bd3ba572e3fcc07f13149839183384a/src/constrainteq.cpp#L1084).

**Comment by dgramop** [comment]
For the next release, should we change the file extension (slvs2)? We can still let the new release open older slvs files, but when saved they'd be converted to slvs2. Parametric constraints will probably also confuse older versions.

**Comment by dgramop** [comment]
Something I noticed is that underscores are interpreted as "operators" which triggers an unknown operator exception. I'm a big fan of snake case, so I think for better user of parameters we should change this in the expr parser

**Comment by ruevs** [comment]
> Something I noticed is that underscores are interpreted as "operators" which triggers an unknown operator exception. I'm a big fan of snake case, so I think for better user of parameters we should change this in the expr parser

@dgramop do you mean this? https://github.com/solvespace/solvespace/blob/7d379e7f3b045ae579537abd392d5b368a4117e5/src/expr.cpp#L692
To me it looks like they are ignored as part of "numbers" here: https://github.com/solvespace/solvespace/blob/7d379e7f3b045ae579537abd392d5b368a4117e5/src/expr.cpp#L693

This is a question to @jwesthues. Jonathan what was your plan idea to ignore `_`? E.g. the following are accepted as "700000" in a distance constraint:
- 700000
- 7E5
- 7e5
- 7__e_5___
- 7_0_0_0_0_0_____

**Comment by dgramop** [comment]
@ruevs while LexNumber allows underscores, when the generic `Lex(...)` function is called, `LexNumber(...)` is only attempted `if(isdigit(c) || c == '.')`. So an underscore outside the context of a number `ispunct` so is attempted to be parsed as an operator.

**Comment by dgramop** [comment]
https://youtu.be/L8cyW_mXisE

There's another bug here with how constraints are taken care of, I haven't really done much with inter-group named parameters. I think taking care of this falls under variable scoping so maybe we visit that in the next PR... Just wanted to document this here so I can pick up after it

**Comment by phkahler** [comment]
> Reference

This is *exactly* the scope issue I talked about previously.  The test_test parameter is first created in the 2D sketch. Later, when it is used in the 3D extrusion, it should be treated as a constant value because it was defined and "solved" in a previous group - so the extrusion height should not be draggable.  This should be equivalent to putting an equal length constraint on the height of the extrusion and the width of the extruded rectangle/square.

AFAICT part of the issue is that the named parameters are created during expression parsing and are not associated with any group.  I'm not sure how such an association is normally created, but it seems to be related to their handles.

One thing I had considered is making a named parameters a "request" that gets displayed in the text window for a group. It should be possible to get the group connection established that way, and we need the functionality anyway. Perhaps adding this capability and creating the parameter prior to use would "fix" the problem. If so, one solution would be to *require* such a definition and NOT create named parameters on the fly during expression evaluation.

TLDR;  Most parameters are created in response to entity creation. Somehow that gets them associated with the group that entity resides in. We need named parameters to have that same group association.

**Comment by ruevs** [comment]
> (Edit: also, how did you get those nice looking links to code with preview?)

1. Browse through the project until you get to the file and line you want:
image
3. Click on the line number.
4. Click on the ellipsis that shows up
6. "Copy permalink"
image

**Comment by ruevs** [comment]
> For the next release, should we change the file extension (slvs2)?

In my opinion - no. We have introduced new features that impact the file format many times since 2.0 (which was the first open source version). Here is a few that I can think of:
- https://github.com/solvespace/solvespace/commit/b23336b5891d4a7b142bde3c9171b100d943da94
- https://github.com/solvespace/solvespace/commit/5df53fc59e7f31e265fabd4c15e6601bd3032833
- https://github.com/solvespace/solvespace/commit/986da7d224319c442dd3600ee7e963315ab3acbb
- https://github.com/solvespace/solvespace/commit/3888909d02b303aab27089588ae78ca41eefb169
- https://github.com/solvespace/solvespace/commit/6245c63f2e1f555f153f1b4dc7c564d139ee5584
- https://github.com/solvespace/solvespace/commit/8cf9d68ecf6304fad9581c076e214b362bfa518a
- https://github.com/solvespace/solvespace/commit/3ccf7845f5271adb9dc097a1cac4f7cfe54f88b3
- https://github.com/solvespace/solvespace/commit/37de36425770880032bd281efe069a1c0b8daea9
- https://github.com/solvespace/solvespace/commit/026f9369894bd526e049ee0fbf2ee928ffdde612

In this cases the old versions simply show the warning - they always do - @jwesthues was forward looking enough to ensure that back in 2009 https://github.com/solvespace/solvespace/commit/b974a4adeb6986febf0d00069f2afe099b753ea1#diff-09121c6f65665acb5c49233fc2f9d269f6da3f97c924fc5ff34aec625f0b01f7R464

So let them show the warning and then "crash and burn" or do whatever the particular new features in the file will cause - many times the entities/constraints are simply missing.

**Comment by ruevs** [comment]
@dgramop do not forget the [other loose end](https://github.com/solvespace/solvespace/pull/1432) that [brought you here](https://github.com/solvespace/solvespace/issues/1428) ;-) Search for `Op::VARIABLE`.
https://github.com/solvespace/solvespace/blob/7d379e7f3b045ae579537abd392d5b368a4117e5/src/expr.h#L22

As I get deeper into the parameters branch I'll also keep it in mind.

**Comment by jwesthues** [comment]
@ruevs I probably intended to allow underscores in literal numbers so that you could write stuff like 1_000_000, inspired by the corresponding behavior in C, and implemented it lazily and half-heartedly. There should be no problem fixing it in the obvious way.

And yeah, no change to the file extension. I'd avoid unnecessary breakage where possible (e.g., don't generate new stuff in the saved file if the new feature isn't actually used), but otherwise just let the users get that warning to upgrade.

**Comment by dgramop** [comment]
@phkahler 
> that the named parameters are created during expression parsing and are not associated with any group. I'm not sure how such an association is normally created, but it seems to be related to their handles.
> making a named parameters a "request" that gets displayed in the text window for a group

I have some code in my other branch that lists parameters by group, but it does so in a roundabout way: by first collecting parameters by constraint in a set, then displaying those parameters. 

> TLDR; Most parameters are created in response to entity creation. Somehow that gets them associated with the group that entity resides in. We need named parameters to have that same group association.

@ruevs 
> We have introduced new features that impact the file format many times since 2.0

Fair, and the way the format is structured if they don't use the RELATION constraint, savefiles will be backwards compatible. I walk back that suggestion

**Comment by dgramop** [comment]
> @dgramop do not forget the [other loose end](https://github.com/solvespace/solvespace/pull/1432) that [brought you here](https://github.com/solvespace/solvespace/issues/1428) ;-) Search for [`Op::VARIABLE`](https://github.com/solvespace/solvespace/blob/7d379e7f3b045ae579537abd392d5b368a4117e5/src/expr.cpp#L720).
> 
> https://github.com/solvespace/solvespace/blob/7d379e7f3b045ae579537abd392d5b368a4117e5/src/expr.h#L22

Wow! That feels like an eternity ago!

**Comment by ruevs** [comment]
> Wow! That feels like an eternity ago!

This is closer to an "eternity ago" ;-)
https://github.com/solvespace/solvespace/commit/61f248899a4f3c4ce06e63759669fe6c3e727417

File: src/mouse.cpp [hunk 2]
```diff
@@ -1429,32 +1465,71 @@ void GraphicsWindow::EditControlDone(const std::string &s) {
...
         if(usedParams > 0) {
-          c->expression = s;
+            c->expr_scaling_to_base = SS.MmPerUnit(); 
+            c->expression = s;
         }
 
         switch(c->type) {
+            case Constraint::Type::RELATION:
+                // on relation, expr_scaling_to_base should be ignored, since the scaling is done on constraints that directly constrain some entity
+                c->expr_scaling_to_base = 1;
```
Comment (ruevs) [inline src/mouse.cpp:1513]: But 1506?

File: src/mouse.cpp [hunk 3]
```diff
@@ -1478,10 +1553,13 @@ void GraphicsWindow::EditControlDone(const std::string &s) {
...
             default:
                 // These are always positive, and they get the units conversion.
-                c->valA = fabs(SS.ExprToMm(e));
+                // Use ExprToMm since this unparameterized expression will simplify immediately, and the dimension transformation is irrelevant
+                c->valA = fabs(SS.NonConstraintExprToMm(e));
                 break;
         }
+        SK.GetGroup(c->group)->dofCheckOk = false; // if an named param was used in the constraint, the DoF may have changed
         SS.MarkGroupDirty(c->group);
+        SS.TW.HideEditControl();
```
Comment (ruevs) [inline src/mouse.cpp:1562]: Why did you have to add this? (I have not tried to remove it and debug what happens).

**Review by ruevs — COMMENTED** [review]
@dgramop I finally found time to review your changes carefully. My comments are mostly minor. Please take a look.

Also you may want to merge my PR.

**Comment by dgramop** [comment]
sorry folks! I got busy with work and was at hackathon but will be reviewing this ASAP.

File: src/drawconstraint.cpp [hunk 4]
```diff
@@ -27,7 +27,10 @@ std::string Constraint::Label() const {
         result = SS.MmToStringSI(fabs(valA));
     }
     if(expression != "") {
-        result = expression;
+        result = expression ;
+        if(SS.MmPerUnit() != expr_scaling_to_base && expr_scaling_to_base != 0) {
```
[the remark this answers is not in the document]
Comment (dgramop) [reply src/drawconstraint.cpp:31]: Will address in next commit, thanks

File: src/expr.cpp [hunk 5]
```diff
@@ -795,8 +797,13 @@ ExprParser::Token ExprParser::Lex(std::string *error) {
 }
 
 ExprParser::Token ExprParser::PopOperand(std::string *error) {
-    Token t = Token::From();
-    if(stack.empty() || stack.back().type != TokenType::OPERAND) {
+    Token t;
```
[the remark this answers is not in the document]
Comment (dgramop) [reply src/expr.cpp:800]: Don't think so, I think I was just explicitly calling constructor for readability. Should I revert this change?

File: src/expr.cpp [hunk 6]
```diff
@@ -806,7 +813,7 @@ ExprParser::Token ExprParser::PopOperand(std::string *error) {
 }
 
 ExprParser::Token ExprParser::PopOperator(std::string *error) {
-    Token t = Token::From();
+    Token t;
```
[the remark this answers is not in the document]
Comment (dgramop) [reply src/expr.cpp:816]: Don't think so, I think I was just explicitly calling constructor for readability. Should I revert this change?

**Comment by dgramop** [comment]
@ruevs addressed some of the items in your review & asked some follow-up questions... will commit review changes soon

**Comment by dgramop** [comment]
@ruevs ping!

File: src/expr.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ruevs) [reply src/expr.cpp]: i FIXED IT NOW In case of error a "proper" `TokenType::ERROR` with `*expr = NULL` is returned. It probably does not matter but this was the old behaviour.

File: src/expr.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (ruevs) [reply src/expr.cpp]: i FIXED IT NOW In case of error a "proper" `TokenType::ERROR` with `*expr = NULL` is returned. It probably does not matter but this was the old behaviour.

**Comment by ruevs** [comment]
Hi @dgramop sorry - somehow I thought you still wanted to change something.

I rebased the branch onto master and merged my pull request about styling `RELATION` constraints. I also changed the two `Token t = Token::From();` places.

With this I am ready and I'll merge this into https://github.com/solvespace/solvespace/tree/parameters.

**Comment by dgramop** [comment]
Use the = button in the toolbar or hit forward slash for the keybind

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
